[8.0] Fix prepared statements on mysql2 adapter

byroot opened the pull request:
Backport of https://github.com/rails/rails/pull/53702, https://github.com/rails/rails/pull/53692, https://github.com/rails/rails/pull/53745 and https://github.com/rails/rails/pull/53799/

FYI @zzak

byroot committed:
[8.0] Fix prepared statements on mysql2 adapter

matthewd reviewed (changes requested):
Let's get it right on main first 😅

byroot commented:
Yeah, the timing is unfortunate. I'll address your review.

byroot committed:
More Mysql2 adapter fixes to support prepared statements

Followup: https://github.com/rails/rails/pull/53702

- Changed Mysql2 type_cast method to also coerce raw Time values.
- Cleaned up some tests.
- Improved the logic around when to skip using a prepared statement.

byroot committed:
Workaround for mysql2 prepared statements flaky test

Ref: https://github.com/brianmario/mysql2/pull/1383

The proper fix can only be upstream, however we can reduce the likeliness
of hitting the bug by only calling `#affected_rows` when strictly necessary,
meaning when we performed a query that didn't return a result set.

Co-Authored-By: Étienne Barrié <etienne.barrie@gmail.com>

byroot committed:
Eagerly close Mysql2 prepared statements

Ref: https://github.com/brianmario/mysql2/pull/1383

If we eagerly close uncached statements, we drastically reduce the
risk of one of them being GCed between `#query` and `#affected_rows`.

The bug is still there thouhg, and can happen when `#execute` is
used.

byroot commented:
Hum, somehow the bug is still happening on the backport? I got a green on `main`. 🤔
